freqtrade/optimize/hyperopt.py

```diff
@@ -328,6 +341,9 @@ def generate_optimizer(self, raw_params: List[Any]) -> Dict[str, Any]:
             self.backtesting.strategy.trailing_only_offset_is_reached = \
                 d['trailing_only_offset_is_reached']
 
+        if HyperoptTools.has_space(self.config, 'trades'):
+            self.max_open_trades = params_dict['max_open_trades']
```

Comment:
are you sure this works throughout the bot in all cases?

the strategy is not reloaded on every hyperopt loop - so the sync between parameter and config will not happen implicitly (and is also not necessary for all other attributes).

`config['max_open_trades']` is used throughout the bot - for example in `wallets.py` to calculate dynamic stakes.
So while this will hyperopt max_open_trades - the stake-amount will remain identical 

example with a 1000 USDT wallet, unlimited stake, and an initial value of 3 (from the config).
with that we get 1000/3 (roughly 333$)$ trades.
If the value is changed to 1 - we'd expect the stake size to be 1000$ (all in on every trade), or a 100$ stake if the trade-amount is bumped to 10.

In my tests, the first stake was always 329.73 - independent of the actually selected trade count - so if i'd start with a 1 max_open_trades value - i'd never get different results - as the stake-amount won't change - implicitly also maxing out all trade slots..

Reply:
> Sorry for replying late

don't think about it that way - that's ... just wrong :)  2 days is NOT late.
it's not our job - so having other priorities in life is perfectly fine. I can't guarantee a review within a day (or whatever), either.

---

i think option 2 will be preferable, at least short-term, otherwise you end up passing `strategy.max_open_trades` around to tons of places.
You'll struggle with option 1 if you attempt to do this "partially" - as there's a ton of "call from a call" scenarios, so even if a function is not directly used in backtesting - it may still be called implicitly (Like the functions in wallets, for example).

in theory, it SHOULD work without leaking to the other epochs in that batch (without copying the config) as hyperopt pickles/unpickles the objects, so it's technically different object id's in each epoch - though i didn't test if the config might leak or not.
it should be easy to test within the strategy with `print(f'{self.max_open_trades}, {self.config["max_open_trades"]}')` or some variation of this ... with the expectation that they both yield the same result every time.

The problem with copying the config is that then you'll need to re-init all objects with the copied config (for example wallets) - which we'd want to avoid if possible.

Reply:
The new changes look to work as expected after a very quick test. :+1: 

i'll try to make some time to test and review properly soon.